### PR TITLE
Support adding a field with ADD COLUMN in Iceberg

### LINKED ISSUE #16248: Allow adding a specific field in nested field in iceberg connector
We're going to support for dropping nested fields in #15975. It would be nice to support adding nested fields as well.

### PR DESCRIPTION
**Pull request by ebyhr**
## Description

Relates to https://github.com/trinodb/trino/discussions/16897
Fixes #16248

## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# General & Iceberg
* Add support for adding nested fields with an `ADD COLUMN` statement. ({issue}`16248`)
```

### CODE REVIEW

File: core/trino-parser/src/main/antlr4/io/trino/sql/parser/SqlBase.g4 [hunk 1]
```diff
@@ -196,7 +196,7 @@ tableElement
     ;
 
 columnDefinition
-    : identifier type (NOT NULL)? (COMMENT string)? (WITH properties)?
+    : qualifiedName type (NOT NULL)? (COMMENT string)? (WITH properties)?
```
[the remark this answers is not in the document]
Comment (findepi) [reply core/trino-parser/src/main/antlr4/io/trino/sql/parser/SqlBase.g4:199]: cc @martint

**Review by findepi — COMMENTED** [review]
"Parse column name as QualifiedName in column definition"

File: core/trino-parser/src/main/antlr4/io/trino/sql/parser/SqlBase.g4 (hunk 1 above)
[the remark this answers is not in the document]
Comment (ebyhr) [reply core/trino-parser/src/main/antlr4/io/trino/sql/parser/SqlBase.g4]: > will we support comment?

I think we won't. Added the below overload:
```
    | ALTER TABLE (IF EXISTS)? tableName=qualifiedName
        ADD COLUMN (IF NOT EXISTS)? columnName=qualifiedName type      #addColumn
```

**Comment by ebyhr** [comment]
Addressed comments.

File: core/trino-main/src/main/java/io/trino/execution/AddColumnTask.java [hunk 2]
```diff
@@ -125,15 +127,82 @@ public ListenableFuture<Void> execute(
...
+            ColumnMetadata column = ColumnMetadata.builder()
+                    .setName(columnName)
+                    .setType(type)
+                    .setNullable(element.isNullable())
+                    .setComment(element.getComment())
+                    .setProperties(columnProperties)
+                    .build();
+            plannerContext.getMetadata().addColumn(session, tableHandle, column);
+        }
+        else {
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/execution/AddColumnTask.java:146]: We have to independent execution paths.
Would it make sense to split `AddColumnTask.java` into `AddColumnTask.java` and `Add[Column]FieldTask.java`?

File: core/trino-parser/src/main/java/io/trino/sql/tree/ColumnDefinition.java [hunk 3]
```diff
@@ -25,26 +25,31 @@
 public final class ColumnDefinition
         extends TableElement
 {
-    private final Identifier name;
+    private final QualifiedName field;
```
Comment (findepi) [inline core/trino-parser/src/main/java/io/trino/sql/tree/ColumnDefinition.java:28]: still needed?

File: core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java [hunk 4]
```diff
@@ -432,6 +432,17 @@ default void addColumn(ConnectorSession session, ConnectorTableHandle tableHandl
         throw new TrinoException(NOT_SUPPORTED, "This connector does not support adding columns");
     }
 
+    /**
+     * Add the specified field, potentially nested, to a row.
+     *
+     * @param parentPath path to a field within the column, without leaf field name.
+     */
+    @Experimental(eta = "2023-06-01") // TODO add support for rows inside arrays and maps and for anonymous row fields
+    default void addField(ConnectorSession session, ConnectorTableHandle tableHandle, List<String> parentPath, String fieldName, Type type)
```
Comment (findepi) [inline core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:441]: do we want to pass `ignoreExisting` into the connector?

(we don't do that for addColumn, but i think we should, since only the connector can do things atomically, engine cannot)

File: testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java [hunk 5]
```diff
@@ -2429,6 +2430,66 @@ protected void verifyAddNotNullColumnToNonEmptyTableFailurePermissible(Throwable
         throw new AssertionError("Unexpected adding not null columns failure", e);
     }
 
+    @Test
+    public void testAddRowField()
+    {
+        if (!hasBehavior(SUPPORTS_ADD_FIELD)) {
+            if (!hasBehavior(SUPPORTS_ADD_COLUMN) || !hasBehavior(SUPPORTS_ROW_TYPE)) {
```
Comment (findepi) [inline testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java:2437]: `!hasBehavior(SUPPORTS_ADD_COLUMN) ||` is probably redundant

File: testing/trino-testing/src/main/java/io/trino/testing/TestingConnectorBehavior.java [hunk 6]
```diff
@@ -67,6 +67,7 @@
 
     SUPPORTS_ADD_COLUMN,
     SUPPORTS_ADD_COLUMN_WITH_COMMENT(SUPPORTS_ADD_COLUMN),
+    SUPPORTS_ADD_FIELD(fallback -> fallback.test(SUPPORTS_ADD_COLUMN) && fallback.test(SUPPORTS_ROW_TYPE)),
```
Comment (findepi) [inline testing/trino-testing/src/main/java/io/trino/testing/TestingConnectorBehavior.java:70]: nice!

File: testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java [hunk 7]
```diff
@@ -2429,6 +2430,66 @@ protected void verifyAddNotNullColumnToNonEmptyTableFailurePermissible(Throwable
         throw new AssertionError("Unexpected adding not null columns failure", e);
     }
 
+    @Test
+    public void testAddRowField()
```
Comment (findepi) [inline testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java:2434]: nit
This is part of "Support adding a field with ADD COLUMN in engine".
At this point, no connector supports the new operation so the test is dead code. You can move BCT to the second commit.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 8]
```diff
@@ -1584,6 +1584,33 @@ public void addColumn(ConnectorSession session, ConnectorTableHandle tableHandle
...
 
+    @Override
+    public void addField(ConnectorSession session, ConnectorTableHandle tableHandle, List<String> parentPath, String fieldName, io.trino.spi.type.Type type)
+    {
+        // Iceberg disallows ambiguous field names in a table. e.g. (a row(b int), "a.b" int)
+        String parentName = String.join(".", parentPath);
+
+        Table icebergTable = catalog.loadTable(session, ((IcebergTableHandle) tableHandle).getSchemaTableName());
+        NestedField parent = icebergTable.schema().findField(parentName);
+        verify(parent.type().isStructType(), "Expect struct type");
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:1595]: You're verifying user input here. When field is not found, we should throw TrinoException

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 9]
```diff
@@ -1584,6 +1584,33 @@ public void addColumn(ConnectorSession session, ConnectorTableHandle tableHandle
...
+        if (parent.type().asStructType().caseInsensitiveField(fieldName) != null) {
+            // TODO https://github.com/trinodb/trino/issues/16273 Adding fields with different cases leads to query failure
+            throw new TrinoException(NOT_SUPPORTED, "This connector does not support adding ambiguous fields");
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:1598]: Do we need to say "This connector" here?
Can this be "Iceberg connector"?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 10]
```diff
@@ -1584,6 +1584,33 @@ public void addColumn(ConnectorSession session, ConnectorTableHandle tableHandle
...
+        }
+
+        // Start explicitly with highestFieldId + 2 to account for existing columns and the new one being
+        // added - instead of relying on addColumn in iceberg library to assign Ids
+        AtomicInteger nextFieldId = new AtomicInteger(icebergTable.schema().highestFieldId() + 2);
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:1603]: Why `+2` and not `+1`?
more importantly, what's wrong with "relying on addColumn in iceberg library to assign Ids"?
basing anything on `highestFieldId` seems wrong, since you can "resurrect" fields that previously had highest numbers and got dropped.

**Review by kasiafi — COMMENTED** [review]
@ebyhr @findepi in my opinion this new functionality is not "ADD COLUMN". It effectively doesn't add a column, but it changes the type of an existing column. Hence, it falls under "ALTER COLUMN".

The user could probably use the existing syntax `ALTER COLUMN ... SET DATA TYPE ...` to add a nested field. I understand however that this syntax is very verbose, as it requires to explicitly describe the requested type. I think that we could extend the `ALTER COLUMN` syntax to support adding a nested field in a more concise form. 
It could be something like:
```
ALTER COLUMN <column_name> ADD FIELD <field_name> <type>
```

The <field_name> could be an identifier for an immediately nested field, and a qualified name for deeper nesting (anyway, the passed name should be relative to the <column_name>).

Execution-wise, the two ways of adding a column, `ALTER COLUMN ... SET DATA TYPE ...` and `ALTER COLUMN ... ADD FIELD ...` should be unified. The new syntax should be considered syntactic sugar.

All the above applies as well to removing nested fields (https://github.com/trinodb/trino/pull/16002).

**Comment by martint** [comment]
Yes, adding nested fields should be handled with SET DATA TYPE. We talked about it nested fields when adding support for that functionality, but stopped before defining the syntax required.

**Comment by findepi** [comment]
> adding nested fields should be handled with SET DATA TYPE.

I agree that adding nested fields should be possible via SET DATA TYPE.

it is, however, impractical to require users to provide current ROW definition plus the new field when they want to add a new field, that's why we need option to add/remove individual fields, or users will not use Trino for that. I don't think there is any disagreement, I just  wanted to make sure others have the context.



> Execution-wise, the two ways of adding a column, `ALTER COLUMN ... SET DATA TYPE ...` and `ALTER COLUMN ... ADD FIELD ...` should be unified. The new syntax should be considered syntactic sugar.

I would be hesitant about that. `ALTER COLUMN ... ADD FIELD ...`  can be executed atomically by a connector, while the equivalent of reading the type, adding a field to the type and invoking `SET DATA TYPE` cannot.

**Comment by findepi** [comment]
> Also, it would be error-prone to maintain two different code paths which can do the same thing

I agree, but it appears a necessity to me. 

A connector can handle operations atomically if and only if it has full knowledge what is the operation (eg "adding a field").
When things are externalized into a task on engine side, there is hardly a way to do this atomically (and still concurrency-friendly).

**Comment by ebyhr** [comment]
CI hit #16652

File: core/trino-main/src/main/java/io/trino/execution/AddFieldTask.java [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+    public AddFieldTask(PlannerContext plannerContext, AccessControl accessControl)
+    {
+        this.plannerContext = requireNonNull(plannerContext, "plannerContext is null");
+        this.accessControl = requireNonNull(accessControl, "accessControl is null");
+    }
+
+    @Override
+    public String getName()
+    {
+        return "ADD COLUMN";
```
Comment (ebyhr) [inline core/trino-main/src/main/java/io/trino/execution/AddFieldTask.java:72]: Should we change to `ADD FIELD`?

**Comment by ebyhr** [comment]
Just rebased on master without any changes.

**Comment by ebyhr** [comment]
@martint @findepi @kasiafi I suppose we already agreed on the syntax in https://github.com/trinodb/trino/issues/16959. The path to array and map types may require further discussion, but those types are excluded in this PR. Could you review this PR when you have time?

File: core/trino-parser/src/main/antlr4/io/trino/sql/parser/SqlBase.g4 [hunk 12]
```diff
@@ -74,6 +74,8 @@ statement
         RENAME TO to=qualifiedName                                     #renameTable
     | ALTER TABLE (IF EXISTS)? tableName=qualifiedName
         ADD COLUMN (IF NOT EXISTS)? column=columnDefinition            #addColumn
+    | ALTER TABLE (IF EXISTS)? tableName=qualifiedName
```
Comment (martint) [inline core/trino-parser/src/main/antlr4/io/trino/sql/parser/SqlBase.g4:77]: This should be part of the existing “add column” syntax (and everything downstream)

**Review by martint — CHANGES_REQUESTED** [review]

File: core/trino-parser/src/main/antlr4/io/trino/sql/parser/SqlBase.g4 (hunk 12 above)
Comment (ebyhr) [reply core/trino-parser/src/main/antlr4/io/trino/sql/parser/SqlBase.g4:77]: @martint It was the initial implementation, but extracted because of https://github.com/trinodb/trino/pull/16321#discussion_r1138841802. Are you suggestion to change `columnDefinition`?

File: core/trino-main/src/main/java/io/trino/execution/AddFieldTask.java [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+        RedirectionAwareTableHandle redirectionAwareTableHandle = plannerContext.getMetadata().getRedirectionAwareTableHandle(session, originalTableName);
+        if (redirectionAwareTableHandle.getTableHandle().isEmpty()) {
+            if (!statement.isTableExists()) {
+                throw semanticException(TABLE_NOT_FOUND, statement, "Table '%s' does not exist", originalTableName);
+            }
+            return immediateVoidFuture();
+        }
+        TableHandle tableHandle = redirectionAwareTableHandle.getTableHandle().get();
+
+        accessControl.checkCanAddColumns(session.toSecurityContext(), redirectionAwareTableHandle.getRedirectedTableName().orElse(originalTableName));
```
Comment (kasiafi) [inline core/trino-main/src/main/java/io/trino/execution/AddFieldTask.java:93]: Should it be `checkCanAddColumns` or `checkCanAlterColumn`?

File: core/trino-main/src/main/java/io/trino/execution/AddFieldTask.java
Comment (kasiafi) [inline core/trino-main/src/main/java/io/trino/execution/AddFieldTask.java:1]: `AddFieldTask` shares a lot of code with `AddColumnTask`. 
If the two use the same grammar, we could unify them also in AST, and in the execution up to some point. I think this is in line with what @martint suggested. We should keep in mind that we're establishing a pattern that we will reuse when adding other field-related features.

One way to unify the grammar for adding columns and fields is to extend `columnDefinition` so that it takes `qualifiedName` instead of `identifier`. That would require validating CREATE TABLE statements in the Analyzer. Additionally, we'd have to filter-out other properties like comment or not null, which are not applicable to nested fields.

**Comment by ebyhr** [comment]
Reverted to the initial approach that changes `columnDefinition` in `SqlBase.g4`.

File: core/trino-main/src/main/java/io/trino/execution/AddColumnTask.java [hunk 14]
```diff
@@ -92,49 +99,121 @@ public ListenableFuture<Void> execute(
...
+
+            ColumnMetadata columnMetadata = plannerContext.getMetadata().getColumnMetadata(session, tableHandle, columnHandles.get(columnName.getValue().toLowerCase(ENGLISH)));
+            Type currentType = columnMetadata.getType();
+            int fieldDepth = 0;
+            for (String fieldName : fieldPath) {
+                fieldDepth++;
+                if (currentType instanceof RowType rowType) {
+                    List<RowType.Field> candidates = rowType.getFields().stream()
+                            // case-sensitive match
+                            .filter(rowField -> rowField.getName().isPresent() && rowField.getName().get().equals(fieldName))
```
Comment (kasiafi) [inline core/trino-main/src/main/java/io/trino/execution/AddColumnTask.java:182]: We should not rely on `Identifier.getName()`. This method breaks the SQL identifier semantics as it ignores quotations.

File: core/trino-main/src/main/java/io/trino/execution/AddColumnTask.java [hunk 15]
```diff
@@ -92,49 +99,121 @@ public ListenableFuture<Void> execute(
...
+                    .collect(toImmutableList());
+
+            ColumnMetadata columnMetadata = plannerContext.getMetadata().getColumnMetadata(session, tableHandle, columnHandles.get(columnName.getValue().toLowerCase(ENGLISH)));
+            Type currentType = columnMetadata.getType();
+            int fieldDepth = 0;
+            for (String fieldName : fieldPath) {
+                fieldDepth++;
+                if (currentType instanceof RowType rowType) {
+                    List<RowType.Field> candidates = rowType.getFields().stream()
+                            // case-sensitive match
```
Comment (kasiafi) [inline core/trino-main/src/main/java/io/trino/execution/AddColumnTask.java:181]: Row field references are resolved case-insensitive in expressions. We should do the same here for consistency.

File: core/trino-main/src/test/java/io/trino/execution/TestAddColumnTask.java [hunk 16]
```diff
@@ -156,18 +161,127 @@ public void testAddColumnOnMaterializedView()
...
+        assertThat(metadata.getTableMetadata(testSession, table).getColumns())
+                .containsExactly(new ColumnMetadata("col", rowType(new RowType.Field(Optional.of("a"), BIGINT))));
+
+        getFutureValue(executeAddColumn(asQualifiedName(tableName), QualifiedName.of("col", "a"), INTEGER, false, true));
+        assertThat(metadata.getTableMetadata(testSession, table).getColumns())
+                .containsExactly(new ColumnMetadata("col", rowType(new RowType.Field(Optional.of("a"), BIGINT))));
+    }
+
+    @Test
+    public void testAddFieldAlreadyExist()
```
Comment (kasiafi) [inline core/trino-main/src/test/java/io/trino/execution/TestAddColumnTask.java:184]: What is the difference between this test method and the following method?

File: core/trino-main/src/main/java/io/trino/execution/AddColumnTask.java (hunk 15 above)
Comment (ebyhr) [reply core/trino-main/src/main/java/io/trino/execution/AddColumnTask.java:181]: How can we handle the table having field paths with different cases? e.g. adding c field to under a.B in `a row(b row(...), B row(...))`

File: core/trino-main/src/main/java/io/trino/execution/AddColumnTask.java (hunk 14 above)
Comment (ebyhr) [reply core/trino-main/src/main/java/io/trino/execution/AddColumnTask.java:182]: The `rowField` comes from the table definition, not comes from user input. I don't think it matters.

File: core/trino-main/src/main/java/io/trino/execution/AddColumnTask.java [hunk 17]
```diff
@@ -92,49 +99,121 @@ public ListenableFuture<Void> execute(
...
+
+            if (!element.isNullable()) {
+                throw semanticException(NOT_SUPPORTED, statement, "Adding fields with NOT NULL constraint is unsupported");
+            }
+            if (!element.getProperties().isEmpty()) {
+                throw semanticException(NOT_SUPPORTED, statement, "Adding fields with column properties is unsupported");
+            }
+            if (element.getComment().isPresent()) {
+                throw semanticException(NOT_SUPPORTED, statement, "Adding fields with COMMENT is unsupported");
+            }
```
[the remark this answers is not in the document]
Comment (ebyhr) [reply core/trino-main/src/main/java/io/trino/execution/AddColumnTask.java:162]: Moved to StatementAnalyzer.

File: core/trino-main/src/main/java/io/trino/execution/AddColumnTask.java (hunk 15 above)
Comment (kasiafi) [reply core/trino-main/src/main/java/io/trino/execution/AddColumnTask.java:172]: If you have a column `a` with type `row(b row(...), B row(...))`, you can't select `a.B`, or any deeper nested field. This is because the `ExpressionAnalyzer` resolves row fields case-insensitive, and finds fields `b` and `B` ambiguous.

My point here is that we should remain consistent in field resolution. That would mean that you cannot add a field under `a.B`. To make sure we use consistent mechanism for field resolution, we could try resolve the `fieldPath` in `StatementAnalyzer`.

Another possible approach would be using the proper SQL identifier semantics. However, there are two issues with this approach:
- the column name is resolved case-insensitive, and only then we switch to another identifier semantics to resolve the remaining path. It might be confusing to the user. The current approach also has this drawback.
- we can canonicalize identifiers on engine side, but then the connectors should be ready to handle them accordingly to the semantics used by the underlying database. This is a complex change, tracked here: https://github.com/trinodb/trino/issues/17

Comment (ebyhr) [reply core/trino-main/src/main/java/io/trino/execution/AddColumnTask.java:172]: > If you have a column a with type row(b row(...), B row(...)), you can't select a.B, or any deeper nested field.

I was missing this fact. Then, case-insensitive makes sense. Thanks for explaining that. I will update this PR.

**Comment by ebyhr** [comment]
Addressed comments.

**Comment by ebyhr** [comment]
Rebased on master to resolve conflicts.

**Comment by ebyhr** [comment]
Rebased on master to resolve conflicts without any changes.

File: core/trino-main/src/test/java/io/trino/execution/TestAddColumnTask.java
Comment (kasiafi) [inline core/trino-main/src/test/java/io/trino/execution/TestAddColumnTask.java:1]: Please add tests for:
- success scenario,
- scenario where on some level of nesting no candidates are found, e.g. adding field `a.b.c` to `row(a row(d))`
- scenario where `not supported` is thrown - i.e. unsupported type found: for structural type, and for "primitive" type

**Review by kasiafi — COMMENTED** [review]
The change looks good overall. However, I'm still concerned about validating the path through resolving it by hand in the `AddColumnTask`. If we want to be able to select the added field, we should apply the same resolution rules as the Analyzer does. We can mimic them for now, but we might run into inconsistency when we start fixing the identifier semantics.

**Comment by ebyhr** [comment]
> We can mimic them for now, but we might run into inconsistency when we start fixing the identifier semantics.

Even if we reuse the existing logic, it will diverge in the future when adding support for array and map types in the field path because we can't use the subscript operators for changing column types.

File: core/trino-main/src/test/java/io/trino/execution/TestAddColumnTask.java
Comment (ebyhr) [reply core/trino-main/src/test/java/io/trino/execution/TestAddColumnTask.java:1]: I didn't add a success scenario intentionally because mock connector doesn't support adding fields and implementing it by ourself doesn't improve the actual test coverage in my opinion.

File: core/trino-main/src/main/java/io/trino/execution/AddColumnTask.java [hunk 18]
```diff
@@ -92,49 +99,111 @@ public ListenableFuture<Void> execute(
...
 
         ColumnDefinition element = statement.getColumn();
+        Identifier columnName = element.getName().getOriginalParts().get(0);
         Type type;
         try {
             type = plannerContext.getTypeManager().getType(toTypeSignature(element.getType()));
         }
         catch (TypeNotFoundException e) {
-            throw semanticException(TYPE_NOT_FOUND, element, "Unknown type '%s' for column '%s'", element.getType(), element.getName());
+            throw semanticException(TYPE_NOT_FOUND, element, "Unknown type '%s' for column '%s'", element.getType(), columnName);
```
[the remark this answers is not in the document]
Comment (ebyhr) [reply core/trino-main/src/main/java/io/trino/execution/AddColumnTask.java:112]: The new `columnName` is an Identifier, not String.

**Comment by ebyhr** [comment]
@martint @kasiafi Please take another look when you have the time.

File: core/trino-main/src/test/java/io/trino/execution/TestAddColumnTask.java [hunk 19]
```diff
@@ -156,18 +164,130 @@ public void testAddColumnOnMaterializedView()
...
+    {
+        QualifiedObjectName tableName = qualifiedObjectName("existing_table");
+        metadata.createTable(
+                testSession,
+                TEST_CATALOG_NAME,
+                rowTable(tableName, new RowType.Field(Optional.of("a"), rowType(new RowType.Field(Optional.of("b"), INTEGER)))),
+                false);
+
+        assertTrinoExceptionThrownBy(() -> getFutureValue(executeAddColumn(asQualifiedName(tableName), QualifiedName.of("col", "x", "c"), INTEGER, false, false)))
+                .hasErrorCode(NOT_SUPPORTED)
```
Comment (kasiafi) [inline core/trino-main/src/test/java/io/trino/execution/TestAddColumnTask.java:197]: It should be `COLUMN_NOT_FOUND`

File: core/trino-main/src/test/java/io/trino/execution/TestAddColumnTask.java [hunk 20]
```diff
@@ -156,18 +164,130 @@ public void testAddColumnOnMaterializedView()
...
+                .hasMessage("Cannot resolve field 'x' within row(a row(b integer)) type when adding [x, c] in row(a row(b integer))");
```
Comment (kasiafi) [inline core/trino-main/src/test/java/io/trino/execution/TestAddColumnTask.java:198]: The added field should render like a dereference expression, not like a list:
`when adding [x, c]` -> `when adding x.c`

**Review by kasiafi — COMMENTED** [review]
>I didn't add a success scenario intentionally because mock connector doesn't support adding fields and implementing it by ourself doesn't improve the actual test coverage in my opinion.

Yes, it should be tested with the target connector.

File: core/trino-main/src/main/java/io/trino/execution/AddColumnTask.java
[the remark this answers is not in the document]
Comment (ebyhr) [reply core/trino-main/src/main/java/io/trino/execution/AddColumnTask.java:1]: I didn't feel much benefit to do it. For instance, dereference logic throws "Column * cannot be resolved" regardless of the type.

**Comment by ebyhr** [comment]
@martint Could you take another look?

**Comment by ebyhr** [comment]
Rebased on master to resolve conflicts.

File: core/trino-main/src/main/java/io/trino/execution/AddColumnTask.java [hunk 21]
```diff
@@ -91,49 +98,111 @@ public ListenableFuture<Void> execute(
...
+                        statement,
+                        "Cannot resolve field '%s' within %s type when adding %s in %s",
+                        fieldName,
+                        currentType,
+                        String.join(".", fieldPath),
+                        columnMetadata.getType());
+            }
+
+            String fieldName = getLast(statement.getColumn().getName().getParts());
+            plannerContext.getMetadata().addField(session, tableHandle, parentPath, fieldName, type, statement.isColumnNotExists());
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/execution/AddColumnTask.java:204]: This is hard to follow due to the many levels of nesting, break/continue, etc.

You can improve it by:
1. Extracting the field resolution to a separate method:

```java
private static List<RowType.Field> getCandidates(RowType rowType, String fieldName)
{
    List<RowType.Field> candidates = rowType.getFields().stream()
            // case-insensitive match
            .filter(rowField -> rowField.getName().isPresent() && rowField.getName().get().equalsIgnoreCase(fieldName))
            .collect(toImmutableList());

    return candidates;
}
```

2. Restructure the loop as follows:

```java
for (int i = 0; i < fieldPath.size() - 1; i++) {
    String fieldName = fieldPath.get(i);

    if (currentType instanceof RowType rowType) {
        List<RowType.Field> candidates = getCandidates(rowType, fieldName);

        if (candidates.isEmpty()) {
            throw semanticException( ... field not found ...);
        }
        if (candidates.size() > 1) {
            throw semanticException(... name is ambiguous ...);
        }
        currentType = getOnlyElement(candidates).getType();
    }
    else {
        throw semanticException( ... field is not of type ROW ...);
    }
}

String fieldName = getLast(statement.getColumn().getName().getParts());
List<RowType.Field> candidates = getCandidates((RowType) currentType, fieldName);

if (!candidates.isEmpty()) {
    if (statement.isColumnNotExists()) {
        return immediateVoidFuture();
    }
    throw semanticException(COLUMN_ALREADY_EXISTS, statement, "Field '%s' already exists", fieldName);
}

...
```

Comment (ebyhr) [reply core/trino-main/src/main/java/io/trino/execution/AddColumnTask.java:204]: Thanks for your suggestion. Updated.

File: core/trino-main/src/main/java/io/trino/execution/AddColumnTask.java [hunk 22]
```diff
@@ -163,47 +164,58 @@ public ListenableFuture<Void> execute(
...
-                        fieldName,
-                        currentType,
-                        String.join(".", fieldPath),
-                        columnMetadata.getType());
             }
 
             String fieldName = getLast(statement.getColumn().getName().getParts());
+            if (!(currentType instanceof RowType)) {
+                throw new TrinoException(NOT_SUPPORTED, "Unsupported type: " + currentType);
+            }
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/execution/AddColumnTask.java:197]: Ah, I missed this in my previous suggestion. Given that you had to add this, it'd be better to move this check to the `getCandidates()` method, update the method to take a Type instead of RowType, remove it from here, and remove the `if (currentType instanceof RowType)` in the loop above.

Comment (ebyhr) [reply core/trino-main/src/main/java/io/trino/execution/AddColumnTask.java]: Updated.

**Comment by ebyhr** [comment]
Rebased on master to resolve conflicts.

**Comment by ebyhr** [comment]
@martint Could you please take another look?

**Comment by findepi** [comment]
@ebyhr the build isn't green

File: testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java [hunk 23]
```diff
@@ -2498,6 +2499,45 @@ protected void verifyAddNotNullColumnToNonEmptyTableFailurePermissible(Throwable
         throw new AssertionError("Unexpected failure when adding not null column", e);
     }
 
+    @Test
+    public void testAddRowField()
+    {
+        if (!hasBehavior(SUPPORTS_ADD_FIELD)) {
+            if (!hasBehavior(SUPPORTS_CREATE_TABLE_WITH_DATA) || !hasBehavior(SUPPORTS_ROW_TYPE)) {
+                return;
```
Comment (findepi) [inline testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java:2507]: `skipTestUnless(hasBehavior(` and move to the top of the method

**Comment by ebyhr** [comment]
The failure #18202 is unrelated to this change.

**Review by findepi — APPROVED** [review]

**Comment by ebyhr** [comment]
Let me merge this PR today if there are no additional comments.

**Comment by martint** [comment]
I’m still reviewing
